ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java

```diff
@@ -0,0 +1,460 @@
+import io.confluent.ksql.integration.Retry;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.rest.client.BasicCredentials;
+import io.confluent.ksql.rest.entity.ActiveStandbyEntity;
+import io.confluent.ksql.rest.entity.ClusterStatusResponse;
+import io.confluent.ksql.rest.entity.KsqlEntity;
+import io.confluent.ksql.rest.entity.KsqlErrorMessage;
+import io.confluent.ksql.rest.entity.KsqlHostInfoEntity;
+import io.confluent.ksql.rest.entity.StreamedRow;
+import io.confluent.ksql.rest.integration.FaultyKafkaConsumer.FaultyKafkaConsumer0;
+import io.confluent.ksql.rest.integration.FaultyKafkaConsumer.FaultyKafkaConsumer1;
+import io.confluent.ksql.rest.integration.FaultyKafkaConsumer.FaultyKafkaConsumer2;
+import io.confluent.ksql.rest.integration.HighAvailabilityTestUtil.Shutoffs;
+import io.confluent.ksql.rest.server.KsqlRestConfig;
+import io.confluent.ksql.rest.server.TestKsqlRestApp;
+import io.confluent.ksql.rest.server.utils.TestUtils;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.PhysicalSchema;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.serde.FormatFactory;
+import io.confluent.ksql.serde.SerdeFeatures;
+import io.confluent.ksql.test.util.KsqlIdentifierTestUtil;
+import io.confluent.ksql.test.util.KsqlTestFolder;
+import io.confluent.ksql.test.util.TestBasicJaasConfig;
+import io.confluent.ksql.util.ConsistencyOffsetVector;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlRequestConfig;
+import io.confluent.ksql.util.UserDataProvider;
+import io.vertx.core.WorkerExecutor;
+import io.vertx.ext.web.RoutingContext;
+import java.io.IOException;
+import java.security.Principal;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import kafka.zookeeper.ZooKeeperClientException;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.streams.StreamsConfig;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test to ensure pull queries route across multiple KSQL nodes correctly.
+ *
+ * <p>For tests on general syntax and handled see RestQueryTranslationTest's
+ * materialized-aggregate-static-queries.json
+ */
+@SuppressWarnings("OptionalGetWithoutIsPresent")
+@Category({IntegrationTest.class})
+@RunWith(MockitoJUnitRunner.class)
+public class PullQueryConsistencyFunctionalTest {
```

Comment:
General comment about this class: There is a bunch of code that is copied from `PullQueryRoutingFunctionalTest`. How about moving the common code to the `HighAvailabilityTestUtil` class?

Reply:
This is objectively good advice.

Just FYI, I could also get behind keeping them independent so that the tests can evolve independently without affecting each other. I think the real deciding factor isn't how much code is similar between the two classes, but how much hassle do you anticipate in the future if we keep them independent or if we create a common util class?

For example, if there's a lot of common code, but it's all pretty straightforward, and you think it'll just sit there not hurting anyone for the next 5 years, then I'd say don't bother combining them. But if it looks like the common code is pretty tricky and you think any change to one test would imply we should also change the other test, then it's better to combine the common code.

In other words, you've got my opinion, and I'm happy whichever way you decide to go.

Reply:
Ok, thanks for both comments @vpapavas @vvcephei . Btw, one reason I didn't extend PullQueryRoutingFunctionalTest in the first place is because all those tests expect 1 partition, and I needed multiple partitions and there is no way to mock the "before" part depending on which tests is executed.

Reply:
@vpapavas I was playing around with moving common parts around, but it's tricky, you move one thing then we need to move another thing et etc, eventually we need to move TestApps and the actual number of it, etc, all things that we might want to change in the future and certain things are all already changed

I think it would be best to keep the copy of those infra calls currently